### PR TITLE
Easy HNF constructors

### PR DESCRIPTION
**Pull request by dellaert**
- I made the HybridNonlinearFactor constructors the same as the earlier PR on HGF.
- I prefer type-checked vectors of shared pointers rather than run-time checks
- that allowed removing templated version and sharing code ebtween two single-key constructors

### CODE REVIEW

File: gtsam/hybrid/HybridNonlinearFactor.cpp [hunk 1]
```diff
@@ -21,10 +21,57 @@
...
+}
+
+/* *******************************************************************************/
+HybridNonlinearFactor::HybridNonlinearFactor(
+    const KeyVector& continuousKeys, const DiscreteKey& discreteKey,
+    const std::vector<NonlinearFactorValuePair>& factors)
+    : Base(continuousKeys, {discreteKey}) {
+  std::vector<NonlinearFactorValuePair> pairs;
+  KeySet continuous_keys_set(continuousKeys.begin(), continuousKeys.end());
+  KeySet factor_keys_set;
```
Comment (varunagrawal) [inline gtsam/hybrid/HybridNonlinearFactor.cpp:62]: Is this not used here?

File: gtsam/hybrid/HybridNonlinearFactor.cpp [hunk 2]
```diff
@@ -21,10 +21,57 @@
...
+  for (auto&& [f, val] : factors) {
+    pairs.emplace_back(f, val);
```
Comment (varunagrawal) [inline gtsam/hybrid/HybridNonlinearFactor.cpp:64]: I don't understand this. Isn't the type of `factors` the same as `pairs`?

File: gtsam/hybrid/HybridNonlinearFactor.h [hunk 3]
```diff
@@ -82,62 +82,49 @@ class GTSAM_EXPORT HybridNonlinearFactor : public HybridFactor {
...
-   * @param keys Vector of keys for continuous factors.
-   * @param discreteKeys Vector of discrete keys.
-   * @param factors Decision tree with of shared factors.
+   * @param continuousKeys Vector of keys for continuous factors.
+   * @param discreteKey The discrete key for the "mode", indexing components.
+   * @param factors Vector of gaussian factors, one for each mode.
    */
-  HybridNonlinearFactor(const KeyVector& keys, const DiscreteKeys& discreteKeys,
-                        const Factors& factors);
+  HybridNonlinearFactor(
```
Comment (varunagrawal) [inline gtsam/hybrid/HybridNonlinearFactor.h:97]: Very nice!

File: gtsam/hybrid/HybridNonlinearFactor.h [hunk 4]
```diff
@@ -196,4 +183,9 @@ class GTSAM_EXPORT HybridNonlinearFactor : public HybridFactor {
       const Values& continuousValues) const;
 };
 
+// traits
+template <>
```
Comment (varunagrawal) [inline gtsam/hybrid/HybridNonlinearFactor.h:187]: Can't believe I forgot about this. Thank you!

File: gtsam/hybrid/tests/Switching.h [hunk 5]
```diff
@@ -161,13 +161,8 @@ struct Switching {
...
       KeyVector keys = {X(k), X(k + 1)};
       auto motion_models = motionModels(k, between_sigma);
-      std::vector<NonlinearFactorValuePair> components;
-      for (auto &&f : motion_models) {
-        components.push_back(
-            {std::dynamic_pointer_cast<NonlinearFactor>(f), 0.0});
-      }
       nonlinearFactorGraph.emplace_shared<HybridNonlinearFactor>(keys, modes[k],
-                                                                 components);
+                                                                 motion_models);
```
Comment (varunagrawal) [inline gtsam/hybrid/tests/Switching.h:165]: I see what you mean by avoiding runtime checks.

**Review by varunagrawal — COMMENTED** [review]
2 minor things

**Comment by dellaert** [comment]
I’ll check comments, but approve? So I can merge both PRs when fixed…

**Review by varunagrawal — APPROVED** [review]

File: gtsam/hybrid/HybridNonlinearFactor.cpp (hunk 1 above)
Comment (dellaert) [reply gtsam/hybrid/HybridNonlinearFactor.cpp:62]: Indeed, removed

File: gtsam/hybrid/HybridNonlinearFactor.cpp (hunk 2 above)
Comment (dellaert) [reply gtsam/hybrid/HybridNonlinearFactor.cpp:64]: Doh, good catch

**Comment by dellaert** [comment]
I addressed (great) comments but will include them  in #1844 to avoid another CI round...

**Pull request merged**